common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -550,6 +558,197 @@ public static UTF8String toTitleCase(final UTF8String target, final int collatio
       BreakIterator.getWordInstance(locale)));
   }
 
+  /**
+   * This 'HashMap' is introduced as a performance speedup. Since titlecasing a codepoint can result
+   * in more than a single codepoint, for correctness, we would use 'UCharacter.toTitleCase(String)'
+   * which returns a 'String'. If we use 'UCharacter.toTitleCase(int)' (the version of the same
+   * function which converts a single codepoint to its titlecase codepoint), it would be faster than
+   * the previously mentioned version, but the problem here is that we don't handle when titlecasing
+   * a codepoint yields more than 1 codepoint. Since there are only 48 codepoints that are mapped to
+   * more than 1 codepoint when titlecased, they are precalculated here, so that the faster function
+   * for titlecasing could be used in combination with this 'HashMap' in the method
+   * 'appendCodepointToTitleCase'.
+   */
+  private static final HashMap<Integer, String> codepointOneToManyTitleCasePrecalculation =
+    new HashMap<>(){{
+    put(223, "Ss");
```

Comment:
Here and below, I would suggest the integer to be written in hex form (specifically as `0xNNNN`). This way if somebody wants to see the corresponding Unicode character, they can easily copy paste the hex form and search using that (essentially the Unicode representation for the character would be U+NNNN which makes it very easy to search for).

Reply:
In case you need more than four hex digits for the integer, just use however many you need (the idea is to use at least four hex digits, not less, and also not more than needed if more than 4).